feat: add nextcloud talk app

CodeShakingSheep opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

This PR adds Nextcloud Talk as conferencing app with dynamic link type to the app store. Makes use of Nextcloud Talk Conversations management API
- [Create a conversation](https://nextcloud-talk.readthedocs.io/en/stable/conversation/#creating-a-new-conversation)
- [Delete a conversation](https://nextcloud-talk.readthedocs.io/en/stable/conversation/#delete-a-conversation)

Fixes #9579
/claim #9579

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). N/A
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Enable Nextcloud Talk in Admin Apps and set the following fields
	- `nextcloudTalkHost`
	- `nextcloundTalkPattern` (optional)
	- `nextcloudTalkUser`
	- `nextcloudTalkPassword`
![image](https://github.com/user-attachments/assets/78ef1e9f-e637-4b73-8658-a39e88964f07)
- Install Nextcloud Talk from App Store
![image](https://github.com/user-attachments/assets/7fa53465-97b5-45e7-8180-cb0e2231be35)
![image](https://github.com/user-attachments/assets/5f1e85d5-6e51-4ddf-a3e2-db76f3dd508d)
- Add Nextcloud Talk to an event type
![image](https://github.com/user-attachments/assets/d505c933-6b0c-4df3-82f0-9d1fb09f6301)
- Create event with Nextcloud Talk as location
![image](https://github.com/user-attachments/assets/7cd40175-5be7-459d-b62a-2606bba2965b)
- Verify in Nextcloud Talk that a new (public) conversation has been created
![image](https://github.com/user-attachments/assets/62d28989-f1a3-4fb8-8f5c-c9e1d48f923a)
- Cancel event with Nextcloud Talk location
![image](https://github.com/user-attachments/assets/b634b18d-72e4-4572-8638-4085dc04366a)
- Verify in Nextcloud Talk that the conversation has been removed

## Checklist

CodeShakingSheep committed:
Initial draft of nextcloud app

CodeShakingSheep committed:
Add video adapter for creating a new conversation via API

CodeShakingSheep committed:
Make links dynamic, add nc user and pwd for auth, fix link creation

CodeShakingSheep committed:
Add nextcloudTalk app to yarn.lock

CodeShakingSheep committed:
Fix screenshot name

anikdhabal commented on yarn.lock:
```diff
@@ -4433,6 +4443,14 @@ __metadata:
   languageName: unknown
   linkType: soft
 
+"@calcom/platform-enums@*, @calcom/platform-enums@workspace:packages/platform/enums":
+  version: 0.0.0-use.local
+  resolution: "@calcom/platform-enums@workspace:packages/platform/enums"
+  dependencies:
+    "@calcom/platform-constants": "*"
+  languageName: unknown
+  linkType: soft
+
```
Great work @CodeShakingSheep , could you pls fix the conflict?

CodeShakingSheep committed:
Merge branch 'main' into feat-add-nextcloud-talk

anikdhabal reviewed (changes requested):
Hey, for a dynamic video app, we need to implement the proper OAuth flow so that users can authorize the app and install it. Currently, I understand from your implementation that it doesn’t follow the proper OAuth process, and I think only an admin account has the ability to use it after changing the keys

CodeShakingSheep commented:
> Hey, for a dynamic video app, we need to implement the proper OAuth flow so that users can authorize the app and install it. Currently, I understand from your implementation that it doesn’t follow the proper OAuth process, and I think only an admin account has the ability to use it after changing the keys

Hey, thanks for the hint. I will have a look at adding OAuth process this week.

Amit91848 reviewed (changes requested):
Thanks for the PR @CodeShakingSheep , but is it possible to do it via OAuth? Also currently individual users cannot make changes to app on their end as keys are only available for admin to edit. That means everyone would be creating meeting on admins nextcloud account

CodeShakingSheep committed:
Add OAuth2 handler

CodeShakingSheep committed:
Fix type for booking reference and fix access token header param, (for creating and deleting a meeting)

CodeShakingSheep commented:
> Thanks for the PR @CodeShakingSheep , but is it possible to do it via OAuth? Also currently individual users cannot make changes to app on their end as keys are only available for admin to edit. That means everyone would be creating meeting on admins nextcloud account

Yes, you're right. It's totally possible to do it via OAuth. NextCloud has an OAuth2 Module. See https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/oauth2.html I just implemented and tested it. It works fine. Feel free to review again @anikdhabal @Amit91848 .

For testing you have to add an OAuth2 client in Nextcloud first (Administration Settings --> Security --> Scroll down to the end). There you need to add the following redirection URL `${WEBAPP_URL}/api/integrations/nextcloudtalk/callback` (I added `http://localhost:3000` for testing purposes). From there you obtain the client ID and client secret.
![image](https://github.com/user-attachments/assets/dda2904d-2036-4137-9b1f-63301e72810c).

In Cal.com in Admin Apps, Nexctloud Talk needs to be enabled with the following fields
- nextcloudTalkHost
- nextcloundTalkPattern (optional)
- nextcloudTalkClientId (obtain from Nextcloud OAuth2 client)
- nextcloudTalkClientSecret (obtain from Nextcloud OAuth2 client)

![image](https://github.com/user-attachments/assets/b8eefffb-3464-4736-aa9c-9d8e7e88108c)

Note for testing: If you had already checked out the branch and activated the Nextcloud Talk app, you need to manually delete the keys from DB `App` table.

CodeShakingSheep committed:
Merge branch 'main' into feat-add-nextcloud-talk

Amit91848 reviewed (commented):
How do you get keys for api? I cannot see administrator tab anywhere

CodeShakingSheep commented:
> How do you get keys for api? I cannot see administrator tab anywhere

Do you mean in Nextcloud or in Cal.com?

**In Nextcloud**
You need to be logged in as admin. Click on your profile in the upper right corner. Then on "Administration settings". Then "Security". On that page scroll down until the end where you see a heading "OAuth 2.0 clients". There you add a client. After adding a new client, you can see `Client Identifier` and `Secret key`.
![image](https://github.com/user-attachments/assets/c74c1615-428e-4fdd-8701-b21dd6ac0349)
Those are the values you need to fill into Cal.com Admin Apps Keys for Nextcloud.

**In Cal.com**
You enter Nextcloud OAuth client's `Client Identifier` in the field `nextcloudTalkClientId` and `Secret key` in `nextcloudTalkClientSecret`.
![image](https://github.com/user-attachments/assets/b8eefffb-3464-4736-aa9c-9d8e7e88108c)

CodeShakingSheep commented on yarn.lock:
```diff
@@ -4777,6 +4786,15 @@ __metadata:
   languageName: unknown
   linkType: soft
 
+"@calcom/synthflow@workspace:packages/app-store/synthflow":
+  version: 0.0.0-use.local
+  resolution: "@calcom/synthflow@workspace:packages/app-store/synthflow"
+  dependencies:
+    "@calcom/lib": "*"
+    "@calcom/types": "*"
+  languageName: unknown
+  linkType: soft
+
```
This came in during fixing a merge conflict https://github.com/calcom/cal.com/pull/16455/commits/5ebda7a6493e57a172d27661a9d8690ddac11e0d#diff-51e4f558fae534656963876761c95b83b6ef5da5103c4adef6768219ed76c2de

And it's in `yarn.lock` file on the main branch https://github.com/calcom/cal.com/blob/main/yarn.lock#L5443. Reverting the whole merge commit is quite some work as a bunch of conflicts would have to be resolved.

CodeShakingSheep committed:
Remove comment

CodeShakingSheep commented on packages/app-store/nextcloudtalk/api/add.ts:
```diff
@@ -0,0 +1,37 @@
+import { encodeOAuthState } from "../../_utils/oauth/encodeOAuthState";
+
+async function handler(req: NextApiRequest) {
+  // Get user
```
:heavy_check_mark: Done

Amit91848 committed:
Merge branch 'main' into feat-add-nextcloud-talk

anikdhabal committed:
Merge branch 'main' into feat-add-nextcloud-talk

anikdhabal commented on packages/app-store/nextcloudtalk/DESCRIPTION.md:
```diff
@@ -0,0 +1,7 @@
+---
+items:
+  - nextcloudtalk1.jpg
+  - nextcloudtalk2.jpg
+---
+
+{DESCRIPTION}
```
Hey, please add a detailed description of how to retrieve the keys that we are adding, step by step. Take our zoom app description as a reference

CodeShakingSheep replied:
Can you elaborate on that? There is no description here about retrieving the keys for Zoom: https://github.com/calcom/cal.com/blob/main/packages/app-store/zoomvideo/DESCRIPTION.md

anikdhabal replied:
Yeah, sorry, it's not in the description, but please create a README file for that Take reference:- https://github.com/calcom/cal.com/blob/main/packages/app-store/webex/README.md

CodeShakingSheep replied:
Just added a README file :heavy_check_mark:

anikdhabal replied:
Thanks @CodeShakingSheep. Lastly, can you provide a complete Loom video where you show everything about the workflow?

CodeShakingSheep replied:
I don't have a Loom account. So, I recorded my screen locally. Here is the link to the video: https://e.pcloud.link/publink/show?code=XZJhYCZE9glp5vsXffQAfvwEmQEK8Urv270
It shows the following:
- Adding an OAuth client in Nextcloud for Cal.com
- Setting up the Nextcloud Talk app in Cal.com (including setup of client ID and secret)
- Activating Nextcloud Talk app for a Cal.com user (including authentication via Nextcloud)
- Setting Nextcloud Talk as location for an event
- Automatic conversation creation in Nextcloud Talk for a booked event
- Automatic conversation removal in Nextcloud Talk for a cancelled event

anikdhabal replied:
Excellent work, @CodeShakingSheep. Is the location link a redirect for conversion? If so, please prioritize the video or meeting link instead. This means that when a booking is made, it should generate a Nextcloud meeting link

CodeShakingSheep replied:
The location link is a direct link to the Nextcloud Talk call, like `<NEXTCLOUD_URL>/call/<ID>` See screenshot from the demo booking I created in the video.
![image](https://github.com/user-attachments/assets/d908f2cf-30b0-4bf8-9a11-b68fd6d2ab16)

CodeShakingSheep committed:
Add README.md file

CodeShakingSheep commented on packages/app-store/nextcloudtalk/lib/VideoApiAdapter.ts:
```diff
@@ -0,0 +1,209 @@
+          grant_type: "refresh_token",
+        };
+        const query = stringify(params);
+        return fetch(`${nextcloudTalkHost}/index.php/apps/oauth2/api/v1/token?${query}`, {
```
Occasionally, while testing I had HTTP 500 responses from my Nextcloud API on the first time of calling this endpoint. When I then called it for the second time (exact same call and params) it just worked. This also happened when I used my local API client for testing which makes me assume that my Nextcloud instance is the issue and not this integration. Also, it only happened sometimes. So, I could not consistently reproduce that behavior.

However, what I did locally was to add a `fetchRetry.js` file in `packages/app-store/_utils` with this code:

```js
async function fetchRetry(input: RequestInfo | URL, init?: RequestInit) {
  let count = 3;
  while (count > 0) {
    try {
      return await fetch(input, init);
    } catch (error) {}

    count -= 1;
  }

  throw new Error(`Too many retries`);
}

export default fetchRetry;
```

So when importing `fetchRetry` this line would become
```suggestion
        return fetchRetry(`${nextcloudTalkHost}/index.php/apps/oauth2/api/v1/token?${query}`, {
```

Like this I never had failed API requests. However, as it's probably just my Nextcloud instance I'm unsure if I should commit this. It's would be just a precaution. What do you think about that @Amit91848 ?

CodeShakingSheep commented on packages/app-store/nextcloudtalk/api/callback.ts:
```diff
@@ -0,0 +1,70 @@
+  const client_id = appKeys.nextcloudTalkClientId as string;
+  const client_secret = appKeys.nextcloudTalkClientSecret as string;
+
+  const result = await fetch(`${hostUrl}/index.php/apps/oauth2/api/v1/token`, {
```
Same as in `VideoApiAdapter.ts`, it might make sense to use `fetchRetry` here instead to deal with HTTP 500 responses on the first call.
```suggestion
  const result = await fetchRetry(`${hostUrl}/index.php/apps/oauth2/api/v1/token`, {
```

anikdhabal replied:
Let's keep it simple for now; we will follow up if we find this for any other user later

CodeShakingSheep committed:
Fix redirect URL for authenticating Nextcloud user

anikdhabal reviewed (approved):
Great Work @CodeShakingSheep , LGTM

PeerRich commented:
@CodeShakingSheep where do I find 

nextcloudTalkHost
nextcloundTalkPattern (optional)
nextcloudTalkUser
nextcloudTalkPassword

?

any link that i can sign up for the credentials?

ad1196 commented:
Where is Nexcloud Talk on online version ? Because not only can I not use the local version due to the persistent localhost issue when deploying on my server, but also the Nextcloud link is empty in the server version of Cal.com. Will the online version ever support Nextcloud?

CodeShakingSheep commented:
> Where is Nexcloud Talk on online version ? Because not only can I not use the local version due to the persistent localhost issue when deploying on my server, but also the Nextcloud link is empty in the server version of Cal.com. Will the online version ever support Nextcloud?

Not sure I understand you correctly. You need to self-host Nextcloud and then enter your Nextcloud base URL as `nextcloudTalkHost` in the Cal.com modal (see comments above).

onerpsystems commented:
I can't find the app in the cal.com app store, what happened?
